crates/core/src/host/wasmtime/mod.rs

```diff
@@ -0,0 +1,228 @@
+use std::borrow::Cow;
+use std::sync::Arc;
+
+use anyhow::Context;
+use once_cell::sync::Lazy;
+use wasmtime::{Engine, Linker, Module, StoreContext, StoreContextMut};
+
+use crate::database_instance_context::DatabaseInstanceContext;
+use crate::error::NodesError;
+use crate::hash::Hash;
+
+mod wasm_instance_env;
+mod wasmtime_module;
+
+use wasmtime_module::WasmtimeModule;
+
+use self::wasm_instance_env::WasmInstanceEnv;
+
+use super::scheduler::Scheduler;
+use super::wasm_common::module_host_actor::InitializationError;
+use super::wasm_common::{abi, module_host_actor::WasmModuleHostActor, ModuleCreationError};
+use super::{EnergyMonitor, EnergyQuanta};
+
+static ENGINE: Lazy<Engine> = Lazy::new(|| {
+    let mut config = wasmtime::Config::new();
+    config
+        .cranelift_opt_level(wasmtime::OptLevel::Speed)
+        .consume_fuel(true)
+        .wasm_backtrace_details(wasmtime::WasmBacktraceDetails::Enable);
+    Engine::new(&config).unwrap()
+});
+
+static LINKER: Lazy<Linker<WasmInstanceEnv>> = Lazy::new(|| {
+    let mut linker = Linker::new(&ENGINE);
+    WasmtimeModule::link_imports(&mut linker).unwrap();
+    linker
+});
+
+pub fn make_actor(
+    dbic: Arc<DatabaseInstanceContext>,
+    module_hash: Hash,
+    program_bytes: &[u8],
+    scheduler: Scheduler,
+    energy_monitor: Arc<dyn EnergyMonitor>,
+) -> Result<impl super::module_host::Module, ModuleCreationError> {
+    let module = Module::new(&ENGINE, program_bytes).map_err(ModuleCreationError::WasmCompileError)?;
+
+    let func_imports = module
+        .imports()
+        .filter(|imp| matches!(imp.ty(), wasmtime::ExternType::Func(_)));
+    let abi = abi::determine_spacetime_abi(func_imports, |imp| imp.module())?;
+
+    if let Some(abi) = abi {
+        abi::verify_supported(WasmtimeModule::IMPLEMENTED_ABI, abi)?;
+    }
+
+    let module = LINKER
+        .instantiate_pre(&module)
+        .map_err(InitializationError::Instantiation)?;
+
+    let module = WasmtimeModule::new(module);
+
+    WasmModuleHostActor::new(dbic, module_hash, module, scheduler, energy_monitor).map_err(Into::into)
+}
+
+#[derive(Debug, thiserror::Error)]
+#[error(transparent)]
+enum WasmError {
+    Db(#[from] NodesError),
+    Wasm(#[from] anyhow::Error),
+}
+
+#[derive(Copy, Clone)]
+struct WasmtimeFuel(u64);
+
+impl WasmtimeFuel {
+    /// 1000 energy quanta == 1 wasmtime fuel unit
+    const QUANTA_MULTIPLIER: i128 = 1_000;
+
+    /// Convert from EnergyQuanta to wasmtime fuel. The second element of the tuple is the remainder
+    /// of quanta in the case that `eq` isn't in u64 range. It can be safely ignored if the supplied
+    /// budget isn't being bookkept.
+    fn from_energy_quanta(eq: EnergyQuanta) -> (Self, EnergyQuanta) {
+        if eq.get() < 0 {
```

Comment:
Negative fuel doesn't exist. My reading of the EnergyQuanta type is that the only reason it supports negative values is because it represents an energy balance (i.e. belonging to an account, that can go into the negative rather than hard cut-off). I was considering making a separate unsigned ReducerBudget type, which I might do as a followup cause I feel like adding negative-value handling everywhere just bc EnergyQuanta is signed is more effort than specifically using a signed integer in the case where we care.

Reply:
My thinking with the remainder was that logically, the energy quanta used for a budget would be like, subtracted from the balance. so:

```
budget = 1000
balance -= budget
budget, remainder = from_energy_quanta(budget)
balance += remainder
```
So therefore if the budget is somehow negative, we wouldn't want to zero out the negative balance, so this behavior ensures everything still works even with the same above pseudocode.

Reply:
I mean if you ignore the second tuple element this is just like, the standard way you'd convert between these 2 units. I feel like by forcing you to ignore it, it makes it explicit that it's lossy and there's a remainder that's not paid attention to. And in any case I think I do wanna do a follow up that splits EnergyQuanta into ReducerBudget & EnergyBalance. u64 is more than enough for a budget for a single reducer.

Reply:
Or even if it doesn't narrow the integer width for ReducerBudget, I wanna make it unsigned.

Reply:
I don't agree it's standard because of the special casing negative values.

Reply:
But it's not special casing - it's impossible to have negative fuel (like, it's literally a u64), so this is the only real way to do it. The `-1e9 eq -> (-1e6 fuel, 0 eq)` alternative you mentioned is impossible.

Reply:
(To be clear I don't think this is a big deal at all)